118-backend-validation-finalized

laithAwad opened the pull request:
This validation will check if the file has the correct ".json" format, it will return an error in the front end if a wrong format file is submitted, it will check that the content of the file has the correct outline, if the types of objects inside the uploaded file are correct, and if the size of the variables arrays and points arrays are of the same length.

ratulkai93 committed:
[118] added basic if-else vaildations

validations added for string type variables

ratulkai93 committed:
[118] added array type checks for validation

ratulkai93 committed:
Merge branch 'develop' into 118]-backend-code-cleanup

laithAwad committed:
Trying to connect to a local DB

Removing unnecessary files which had bugs when used on mac when trying to connect to a DB locally created, also commented out some part of the code in fileUploadService.ts that was causing an error and making the creation of new DB get stuck.

laithAwad committed:
Json file upload validation

Added type validation for all sections of the json file, and added error handling for all await functions. This commit also has 2 sample json files for testing purposes.

ratulkai93 committed:
[118]-updated frontend validation for file extension check

the system now displays an error when a file with any extension other than .json is uploaded

laithAwad committed:
Validation for extension to be .json

We have added a validation to check if the file's extension is .json before uploading it to the database, this was done in the frontend part to prevent any errors in the backend as well. Note that importing materialUi is needed for the way we handled this issue and using "npm install @material-ui/lab" to install it on the project folder directory is also needed.

ratulkai93 committed:
[118[ fixed unused datasetcontroller import and wait-for-it file line sequence

modified loadStartupProcess and wait-for-it files to facilitate loading the website

ratulkai93 committed:
[118] added json file content check helper function

this function will allow us to verify if the json item is valid

ratulkai93 committed:
[118] added json file outline validation

the jsonvalidator runs  validity check on the items of the uploaded file content before error handling and type checking

laithAwad committed:
Finalized the validation

Added a few comments and the dependency installation needed to run the files correctly in the comments on top. Removed unnecessary and unused code.

GD-Dheer commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -17,6 +23,17 @@ export default function DataCell() {
         e.preventDefault();
 
         const json = e.target.jsonFile.files[0];
```
The idea for this was not necessarily strict the upload to JSON but instead just validate that the file is an accepted type. In the future if we want to add csv or txt support then this will have to be changed

laithAwad replied:
Correct, the way we handled it will allow us to check any type we decide to add in the future because it is dependant on the extension of the file.

GD-Dheer commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -25,24 +42,47 @@ export default function DataCell() {
+                    <form onSubmit={handleSubmit}>
+                        <img src={require('./uploadimage.png')} alt="Visual of clouds"></img>
+                        <div>
+                            <input type="file" id="jsonFile" accept="application/json" />
```
You will get merge conflict here for id= - lets keep the "file" one currently in develop

laithAwad replied:
I have just fixed all merge conflicts with develop

GD-Dheer commented on JsonExample-2.json:
```diff
@@ -0,0 +1,240 @@
+{
+  "reference": {
```
We dont need this file in our repo

GD-Dheer commented on JsonExample.json:
```diff
@@ -0,0 +1,242 @@
+{
```
Same Comment.` Dont need to keep

GD-Dheer replied:
Same Comment.` Dont need to keep

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
-}
\ No newline at end of file
+
+//Running and installing [npm i jsonschema] is required for this page to work
```
This file needs to look like Typescript. Take a look at this file under develop

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+const uploadModel = require('../models/DataUploadModel');
+//const jsValidator= require('../services/validationSchema');
+
+var Validator = require('jsonschema').Validator;
```
vars -> lets or const where applicable

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+  }
+};
+
+var m2schema = {
```
Can we rename this variable to something more descriptive than m2?

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+  let unitsID = '';
+  let reprID = '';
+
+  //to run validation of the json file outline
```
We should move the validation functions into a separate file

ratulkai93 replied:
the validations have been moved to validationSchema.ts and is properly now imported to fileUploadService.ts

GD-Dheer commented on Server/src/services/validationSchema.ts:
```diff
@@ -0,0 +1,113 @@
+//Running and installing [npm i jsonschema] is required for this page to work
+//More details and examples about json schema validator can be found at:
+// https://github.com/tdegrunt/jsonschema#readme
+
```
Is this duplicated code?

laithAwad replied:
No, this is the instructions to make sure the json schema validator is installed before you run the file, "npm i jsonschema" is required to run it correctly

ds161833 replied:
remove these comments, running npm install is a norm when developing in react/nodejs, packages are being added all the time. and 100% of the time if you get an error, you can see that ur missing a package, and you can immidiately npm install

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+  }
+};
+
+var v2Schema = {
```
Rename this

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+
+};
+
+var referenceSchema = {
```
I would actually rename them all like referencesValidationSchema

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+  }
+};
+
+var c2Schema = {
```
rename

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+  processUpload
+}
+
+const stringValidation = (reference) => {
```
Lets move all validations in another folder and import

GD-Dheer commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -17,6 +23,17 @@ export default function DataCell() {
+            //so if this here is not proper json extension it will catch
+            var ext = getExtension(json.name);
+            if (ext != '.json') {
+                throw "File extension is not supported";
```
Does this work? I dont know if you can just throw as such without indicating the type of error

ratulkai93 replied:
this works, but now i have changed it console.error('file extension is not supported') to avoid confusion

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,118 +1,380 @@
+import { json } from "express";
+const fileSystem = require('fs');
+const uploadModel = require('../models/DataUploadModel');
+//const jsValidator= require('../services/validationSchema');
```
Why is this commented out?

GD-Dheer reviewed (changes requested):
Could you add in the description on github how your schemas work exactly? Also error handling was part of this I believe. Did you guys have time to work on that as well?

Also a handleful of merge conflicts

laithAwad committed:
Final version of file upload

Tested and verified the json schema is working, added final comments to the files. Fixed merge conflicts with develop branch.

laithAwad committed:
Merge branch 'develop' into 118]-backend-code-cleanup

laithAwad committed:
Removed json example files

Removed the files that were used to test the json upload functionality

laithAwad committed:
Typescript compatibility

Made sure the files are converted to typescript

laithAwad committed:
Created ValidationSchema file

Created a file that contains all the schemas to be used by our validator in the fileUploadService

laithAwad committed:
Added ValidationSchema file

This file contains all the schemas to be used by the validator inside the fileUploadService file

ds161833 commented on .gitignore:
```diff
@@ -137,3 +137,7 @@ sketch
+
+Client/package-lock.json
+
+Client/package.json
```
we should not ignore package and package-lock files, npm uses package to check what packages are missing in the current environment, and package lock is there to ensure we got correct dependencies. We should be pushing package lock and package.json files

ds161833 commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -1,47 +1,89 @@
+            console.log("was not able to validate json of uploaded file")
+            setOpen(true)
+            return;
+        }
```
fix indentation in this file

ds161833 commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -1,47 +1,89 @@
+            //so if this here is not proper json extension it will catch
+            var ext = getExtension(json.name);
+            if (ext != '.json') {
+                throw "File extension is not supported";
```
its up to you to throw, I dont mind, but exceptions are meant to cover something that is not in the developer's control, here however instead of throwing we could setOpen(true) and return

ratulkai93 replied:
i have changed the throw to a console.error() instead to avoid confusion for the time being.
just to clarify, what should we try to return?

ratulkai93 replied:
for now, file extension check has been moved out of try/catch block

ds161833 commented on Server/src/services/validationSchema.ts:
```diff
@@ -0,0 +1,128 @@
+        "required": ["authors"]
+    },
+
+    'authorsValidationSchema': {
```
maybe for the future, im sure there is a nice way to achieve this with react instead of hardcoding schema, our models may change, and then this schema too, but we will forget to change it and it will lead to a bug

ratulkai93 replied:
comments have been removed from the files

ratulkai93 replied:
the schema was based on the contents of the json example file that we have been using for the file upload testing. will look into optimize the schemas for future sprints

ds161833 commented on Client/src/Components/DataCell/DataCell.js:
```diff
@@ -1,47 +1,89 @@
-  )
-}
+//Verify that material-ui lab and material-ui core are installed before running this file!
+// npm install @material-ui/lab
```
again this should not be here, 'npm install' installs all the packages inside the package.json in the case they are not installed already, we do not need to leave comments about this. Just run 'npm install' every time you pull develop

ds161833 replied:
consult the docs: 
![image](https://user-images.githubusercontent.com/22376976/101484107-1e500300-3927-11eb-8317-910112d13e3c.png)

ratulkai93 committed:
[118] removed additional comments

ratulkai93 committed:
[118] fixed indentation and changed error handling format

after checking uploaded file extenstion, it will show the error message on the console log instead of using throw statement

ratulkai93 committed:
Merge remote-tracking branch 'origin/develop' into 118]-backend-code-cleanup

ratulkai93 committed:
[118] package.json removed from gitignore and adjuststed file extension verification

removed package.json from gitignore and removed the try catch block from file extention verification on frontend

ds161833 reviewed (approved):
looks good now 👍

GD-Dheer commented on Server/tmp/upload/55b62868cf4cf59724816bf19843575e:
```diff
@@ -0,0 +1,242 @@
+{
+  "reference": {
```
We do not need these files. They created if you test an upload. Avoid committing these to the repo please

ratulkai93 replied:
tmp folder has been deleted/.

GD-Dheer commented on Server/tmp/upload/b4c248d49de808448540ec3a20a58ee3:
```diff
@@ -0,0 +1,242 @@
+{
+  "reference": {
```
Same as above

ratulkai93 replied:
deleted

GD-Dheer reviewed (approved):
Rest looks good

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,132 +1,279 @@
+    let datasetID: number;
+    let unitsID: number;
+    let reprID: number;
+
```
Can we try not to leave so much space?

GD-Dheer commented on Server/src/services/fileUploadService.ts:
```diff
@@ -1,132 +1,279 @@
+    v.addSchema(schemaValidator.variableDetailsValidationSchema, '/v2Schema');
+    v.addSchema(schemaValidator.contentsValidationSchema, '/contentsSchema');
+    v.addSchema(schemaValidator.contentsDetailsValidationSchema, '/c2Schema');
+
```
Does this show as extra spacing?

ratulkai93 committed:
[118] deleted unnecessary json files

unused files generated during upload process have been removed

ratulkai93 committed:
[118] minor spacing fix

ratulkai93 committed:
Merge remote-tracking branch 'origin/develop' into 118]-backend-code-cleanup

ratulkai93 committed:
Merge remote-tracking branch 'origin/develop' into 118]-backend-code-cleanup
